security/src/main/scala/pl/touk/nussknacker/ui/security/api/AuthenticationManager.scala

```diff
@@ -0,0 +1,107 @@
+import cats.data.OptionT
+import pl.touk.nussknacker.security.AuthCredentials.{
+  ImpersonatedAuthCredentials,
+  NoCredentialsProvided,
+  PassedAuthCredentials
+}
+import pl.touk.nussknacker.security.{AuthCredentials, ImpersonatedUserIdentity}
+import pl.touk.nussknacker.ui.security.api.AuthenticationManager.{
+  impersonateHeaderName,
+  impersonationHeaderEndpointInput,
+  mappedAuthenticationEndpointInput
+}
+import sttp.tapir._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuthenticationManager(authenticationResources: AuthenticationResources)(implicit ec: ExecutionContext) {
+
+  lazy val authenticationRules: List[AuthenticationConfiguration.ConfigRule] =
+    authenticationResources.configuration.rules
+  lazy val isAdminImpersonationPossible: Boolean = authenticationResources.configuration.isAdminImpersonationPossible
+
+  def authenticate(): Directive1[AuthenticatedUser] = authenticationResources.getAnonymousRole match {
+    case Some(role) =>
+      authenticationResources.authenticate().optional.flatMap {
+        case Some(user) => provideOrImpersonateUser(user)
+        case None => handleAuthorizationFailedRejection.tmap(_ => AuthenticatedUser.createAnonymousUser(Set(role)))
+      }
+    case None => authenticationResources.authenticate().flatMap(provideOrImpersonateUser)
+  }
+
+  def authenticate(authCredentials: AuthCredentials): Future[Option[AuthenticatedUser]] = authCredentials match {
+    case passedCredentials @ PassedAuthCredentials(_) => authenticationResources.authenticate(passedCredentials)
+    case NoCredentialsProvided                        => authenticateWithAnonymousAccess()
+    case ImpersonatedAuthCredentials(impersonatingUserAuthCredentials, impersonatedUserIdentity) =>
+      authenticateWithImpersonation(impersonatingUserAuthCredentials, impersonatedUserIdentity)
+  }
+
+  def authenticationEndpointInput(): EndpointInput[AuthCredentials] =
+    authenticationResources
+      .authenticationMethod()
+      .and(impersonationHeaderEndpointInput)
+      .map(mappedAuthenticationEndpointInput)
+
+  private def provideOrImpersonateUser(user: AuthenticatedUser): Directive1[AuthenticatedUser] = {
+    optionalHeaderValueByName(impersonateHeaderName).flatMap {
+      case Some(impersonatedUserIdentity) =>
+        authenticationResources.getImpersonatedUserData(impersonatedUserIdentity) match {
+          case Some(impersonatedUserData) =>
+            provide(AuthenticatedUser.createImpersonatedUser(user, impersonatedUserData))
+          case None => reject
+        }
+      case None => provide(user)
+    }
+  }
+
+  private def authenticateWithImpersonation(
+      impersonatingUserAuthCredentials: PassedAuthCredentials,
+      impersonatedUserIdentity: ImpersonatedUserIdentity,
+  ): Future[Option[AuthenticatedUser]] = (for {
+    impersonatingUser <- OptionT(authenticationResources.authenticate(impersonatingUserAuthCredentials))
+    impersonatedUserData <- OptionT(
```

Comment:
`OptionT` has many helpful factory methods.
E.g here we could do it like this:
```scala
impersonatedUserData <- OptionT.fromOption[Future](authenticationResources.getImpersonatedUserData(impersonatedUserIdentity.value))
```

Reply:
Oh, yes it's better!